autoretrieve/engine.go

```diff
@@ -0,0 +1,524 @@
+package autoretrieve
+
+import (
+	"context"
+	"encoding/base64"
+	"fmt"
+	"sync"
+
+	"github.com/filecoin-project/go-legs"
+	"github.com/filecoin-project/go-legs/dtsync"
+	"github.com/filecoin-project/go-legs/httpsync"
+	provider "github.com/filecoin-project/index-provider"
+	"github.com/filecoin-project/index-provider/engine/chunker"
+	"github.com/filecoin-project/index-provider/metadata"
+	"github.com/filecoin-project/storetheindex/api/v0/ingest/schema"
+	"github.com/hashicorp/go-multierror"
+	"github.com/ipfs/go-cid"
+	"github.com/ipfs/go-datastore"
+	dsn "github.com/ipfs/go-datastore/namespace"
+	logging "github.com/ipfs/go-log/v2"
+	"github.com/ipld/go-ipld-prime"
+	cidlink "github.com/ipld/go-ipld-prime/linking/cid"
+)
+
+const (
+	keyToCidMapPrefix      = "map/keyCid/"
+	cidToKeyMapPrefix      = "map/cidKey/"
+	keyToMetadataMapPrefix = "map/keyMD/"
+	latestAdvKey           = "sync/adv/"
+	linksCachePath         = "/cache/links"
+)
+
+var (
+	log = logging.Logger("provider/engine")
+
+	dsLatestAdvKey = datastore.NewKey(latestAdvKey)
+)
+
+// Engine is an implementation of the core reference provider interface
+type AutoretrieveEngine struct {
+	*options
+	lsys ipld.LinkSystem
+
+	entriesChunker *chunker.CachedEntriesChunker
+
+	publisher legs.Publisher
+
+	mhLister provider.MultihashLister
+	cblk     sync.Mutex
+}
+
+// var _ provider.Interface = (*AutoretrieveEngine)(nil)
+
+// New creates a new index provider Engine as the default implementation of provider.Interface.
+// It provides the ability to advertise the availability of a list of multihashes associated to
+// a context ID as a chain of linked advertisements as defined by the indexer node protocol implemented by "storetheindex".
+// Engine internally uses "go-legs", a protocol for propagating and synchronizing changes an IPLD DAG, to publish advertisements.
+// See:
+//  - https://github.com/filecoin-project/storetheindex
+//  - https://github.com/filecoin-project/go-legs
+//
+// Published advertisements are signed using the given private key.
+// The retAddrs corresponds to the endpoints at which the data block associated to the advertised
+// multihashes can be retrieved.
+// Note that if no retAddrs is specified the listen addresses of the given libp2p host are used.
+//
+// The engine also provides the ability to generate advertisements via Engine.NotifyPut and
+// Engine.NotifyRemove as long as a provider.MultihashLister is registered.
+// See: provider.MultihashLister, Engine.RegisterMultihashLister.
+//
+// The engine must be started via Engine.Start before use and discarded via Engine.Shutdown when no longer needed.
+// See: Engine.Start, Engine.Shutdown.
+func New(o ...Option) (*AutoretrieveEngine, error) {
+	opts, err := newOptions(o...)
+	if err != nil {
+		return nil, err
+	}
+
+	e := &AutoretrieveEngine{
+		options: opts,
+	}
+
+	e.lsys = e.mkLinkSystem()
+
+	return e, nil
```

Comment:
given `mhlister` is required, you should validate it is set before creating an engine

Reply:
`mhlister` is not required. You can theoretically use the engine without it.

Reply:
it is required, you cant make the `mkLinkSystem` without it, it will fail. The reference repo is too relaxed, we can improve our implementation, to make sure the engine won't start if `mhlister` is not set - so we don't fail later after main.go bootstraps